Converted the page Platforms/Amazon Web Services into blog post (#468)
etcd-io/website: content/en/blog/2021/etcd-deployments-on-AWS.md

---
title: Deployment of etcd on AWS EC2

> author: Caleb Miles, Somoshree Datta

Reviewer: As with @spzala's comment https://github.com/etcd-io/website/pull/469#discussion_r714322007, probably we should only have Caleb Mile's name here as an author. Not to discount the work you've done on this @Somoshree, but this work is more adaptation than authorship.

I'm not sure we need any authorship in this line as we're already citing Caleb in the note at the bottom of the page.

## Self-healing

While using an ASG to scale the size of an etcd cluster is not recommended, an ASG can be used effectively to maintain the desired number of nodes in the event of node failure. The maintenance of a stable number of etcd nodes will provide the etcd cluster with a measure of self-healing.

## Next steps

The operational life cycle of an etcd cluster can be greatly simplified through the use of the etcd-operator. The open source etcd operator is a Kubernetes control plane operator which deploys and manages etcd clusters atop Kubernetes. While still in its early stages the etcd-operator already offers periodic backups to S3, detection and replacement of failed nodes, and automated disaster recovery from backups in the event of permanent quorum loss.

## **Note** 

> This blog post is an adaptation from a page in the documentation, previously under [Platforms](../../../docs/v3.5/platforms/), which describes etcd deployments on various platform services. The original page in the documention was authored by **Caleb Miles**. As this article is more about etcd deployment on AWS and not really about the core etcd concepts, hence maintainers of this website found it most suitable to remove this page from the actual documentation and put it in as a blog post.

Reviewer: A suggestion for this final note:
```suggestion
This post is an adaptation of a page previously found in the Platforms section of the documentation which described etcd deployments on various platform services. The original page was authored by [**Caleb Miles** and others](https://github.com/etcd-io/etcd/commits/6c08437ec330f84f78a59f7772884c7ef5374247/Documentation/platforms/aws.md). As this page was more about etcd deployment on AWS than core etcd concepts, the etcd maintainers suggested removing the page from the documentation and converting it to this blog post.
```

Removing the link for Platforms, as eventually that section will be removed. Adding a link to the original file's etcd-io/etcd history to illustrate contributions.

---
title: Deployment of etcd on AWS EC2
author: Caleb Miles, Somoshree Datta

> date: 2021-09-03

Reviewer: Do we want to aim for a mid month release of this post? Maybe Sept 14th or 15th? @spzala thoughts?

I think we should release the other two in 2-3 week intervals, #469 maybe in early Oct, and #471 in late Oct or early Nov?

Reviewer 2: @nate-double-u since the post is an old one, I am not sure if we want to put the latest date here. The content is great as we already have it but not sure if anything is changed and putting the current date makes it feel like it's the latest blog and tested accordingly.

Reviewer: That's a good point. @Okeanos makes a similar one over on the [Platforms/Container-Linux-with-systemd post](https://github.com/etcd-io/website/pull/469#issuecomment-922461700).

I think we should back date this as we are providing this info mostly as an archival reference to a page we're removing.

Which date should we use? The original page creation date? the date it was moved? other?

* [Cluster Design](#cluster-design)
* [Availability](#availability)
* [Data durability after member failure](#data-durability-after-member-failure)
* [Perfomance/Throughput](#performancethroughput)
* [Network](#network)
* [Disk](#disk)
* [Self-healing](#self-healing)
* [Next Steps](#next-steps)

## Capacity Planning

As a critical building block for distributed systems it is crucial to perform adequate capacity planning in order to support the intended cluster workload. As etcd is a highly available and strongly consistent data store, so increasing the number of nodes in an etcd cluster will generally affect performance adversely. This makes sense intuitively, as more nodes means more members for the leader to coordinate state across. The most direct way to increase throughput and decrease latency of an etcd cluster is allocate more disk I/O, network I/O, CPU, and memory to the cluster members. In this scenario, it is impossible to temporarily divert incoming requests to the cluster, though scaling the EC2 instances which comprise the etcd cluster members one at a time may improve performance. It is, however, best to avoid bottlenecks through capacity planning.

> The etcd team has produced a [hardware recommendation guide](https://etcd.io/docs/v3.5/op-guide/hardware/) which is very useful for "ballparking" how many nodes and what instance type are necessary for a cluster.

Reviewer: It's a bit more unweildly, but since his is an internal links to the site, I'd recommend using a relative link here.
```suggestion
The etcd team has produced a [hardware recommendation guide](../../../docs/v3.5/op-guide/hardware/) which is very useful for "ballparking" how many nodes and what instance type are necessary for a cluster.
```

AWS provides a service for creating groups of EC2 instances which are dynamically sized to match load on the instances. Using an Auto Scaling Group ([ASG](http://docs.aws.amazon.com/autoscaling/latest/userguide/AutoScalingGroup.html)) to dynamically scale an etcd cluster, is not recommended for several reasons including the following:

* etcd performance is generally inversely proportional to the number of members in a cluster due to the synchronous replication which provides strong consistency of data stored in etcd.

> * The operational complexity of adding [lifecycle hooks](http://docs.aws.amazon.com/autoscaling/latest/userguide/lifecycle-hooks.html) to properly add and remove members from an etcd cluster by modifying the [runtime configuration](https://etcd.io/docs/v3.5/op-guide/runtime-configuration/).

Reviewer: Same as above relative link
```suggestion
* The operational complexity of adding [lifecycle hooks](http://docs.aws.amazon.com/autoscaling/latest/userguide/lifecycle-hooks.html) to properly add and remove members from an etcd cluster by modifying the [runtime configuration](../../../docs/v3.5/op-guide/runtime-configuration/).
```

* **Replication**: etcd replicates all data to all members of the etcd cluster. Therefore, given more members in the cluster and more independent failure domains, the more less-likely it is that the data stored in an etcd cluster will be permanently lost in the event of a disaster.

* **Point in time etcd snapshotting**: The etcd v3 API introduced support for snapshotting clusters. The operation is cheap enough (completing in the order of minutes) to run quite frequently and the resulting archives can be archived in a storage service like Amazon Simple Storage Service (S3).

* **Amazon Elastic Block Storage (EBS)**: An EBS volume is a replicated, network attached block device which have stronger storage safety guarantees than InstanceStore, which in turn, has a life cycle associated with the life cycle of the attached EC2 instance. The life cycle of an EBS volume is not necessarily tied to an EC2 instance and can be detached and snapshotted independently which means that a single node etcd cluster backed by an EBS volume can provide a fairly reasonable level of data durability.

## Performance/Throughput

> The performance of an etcd cluster is roughly quantifiable through latency and throughput metrics which are primarily affected by disk and network performance. Detailed performance planning information is provided in the [performance section](https://etcd.io/docs/v3.5/op-guide/performance/) of the etcd operations guide.

Reviewer: Same as above, relative link
```suggestion
The performance of an etcd cluster is roughly quantifiable through latency and throughput metrics which are primarily affected by disk and network performance. Detailed performance planning information is provided in the [performance section](../../../docs/v3.5/op-guide/performance/) of the etcd operations guide.
```
